### PR TITLE
ARROW-6420: [Java] Improve the performance of UnionVector when getting underlying vectors

### PR DESCRIPTION
**Pull request by liyafan82**
Getting the underlying vector is a frequent opertation for UnionVector. It relies on this operation to get/set data at each index.

The current implementation is inefficient. In particular, it first gets the minor type at the given index, and then compares it against all possible minor types in a switch statment, until a match is found.

We improve the performance by storing the internal vectors in an array, whose index is the ordinal of the minor type. So given a minor type, its corresponding underlying vector can be obtained in O(1) time.

It should be noted that this technique is also applicable to UnionReader and UnionWriter, and support for UnionReader is already implemented.

### CODE REVIEW

**Review by emkornfield — CHANGES_REQUESTED** [review]
Usually switches act as [jumps](https://www.geeksforgeeks.org/switch-vs-else/) not scans.

The change in writers actually can make performance worse because it needs to scan through the entire array instead of just active writers.

I'm skeptical is worth it, but, at the very least I think we need a benchmark to understand how much this does help.

**Comment by liyafan82** [comment]
> Usually switches act as [jumps](https://www.geeksforgeeks.org/switch-vs-else/) not scans.
> 
> The change in writers actually can make performance worse because it needs to scan through the entire array instead of just active writers.
> 
> I'm skeptical is worth it, but, at the very least I think we need a benchmark to understand how much this does help.

Sure I will provide performance benchmark to verify it. 

It should be noted that the proposed method does not scan through the entire array. It just calculates the index and access the element directly. (Maybe I don't fully understand your point?)

**Comment by emkornfield** [comment]
> It should be noted that the proposed method does not scan through the entire array. 

I was referring to setPosition.

**Comment by liyafan82** [comment]
> > It should be noted that the proposed method does not scan through the entire array.
> 
> I was referring to setPosition.

You mean the extra if statement could make the setPosition method slower, right?

**Comment by liyafan82** [comment]
@emkornfield I have provided the benchmark as you suggested.
Results show that by checking and returning non-null vectors at the beginning, the performance improves 2+ times:

Before:
UnionBenchmarks.getObjectBenchmark  avgt    5  34.578 ± 0.349  ns/op

After:
UnionBenchmarks.getObjectBenchmark  avgt    5  16.556 ± 0.577  ns/op

**Comment by liyafan82** [comment]
@emkornfield It is interesting to note that, JIT translated switch statement as a number of jumps, so it also amounts to a series of if statements:

image

File: java/performance/src/test/java/org/apache/arrow/vector/complex/UnionBenchmarks.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  public void tearDown() {
+    vector.close();
+    allocator.close();
+  }
+
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public Object getObjectBenchmark() {
+    return vector.getObject(0);
```
Comment (emkornfield) [inline java/performance/src/test/java/org/apache/arrow/vector/complex/UnionBenchmarks.java]: why getObject instead of getVector?

**Comment by emkornfield** [comment]
Thanks @liyafan82 for the benchmark.  This is a little surprising, I'm still not sure about the change.  CC @jacques-n for his input.

> You mean the extra if statement could make the setPosition method slower, right?
If I understand the code correctly its not the if-statement.  With the next code every MinorType needs to be iterated over instead of just those added to the writer.

File: java/performance/src/test/java/org/apache/arrow/vector/complex/UnionBenchmarks.java (hunk 1 above)
Comment (liyafan82) [reply java/performance/src/test/java/org/apache/arrow/vector/complex/UnionBenchmarks.java]: getVector is not public in the current code base

**Comment by liyafan82** [comment]
> Thanks @liyafan82 for the benchmark. This is a little surprising, I'm still not sure about the change. CC @jacques-n for his input.
> 
> > You mean the extra if statement could make the setPosition method slower, right?
> 
> If I understand the code correctly its not the if-statement. With the next code every MinorType needs to be iterated over instead of just those added to the writer.

@emkornfield Let's discuss over the generated code shown below:

BaseWriter getWriter(MinorType minorType) {
    if (writers[minorType.ordinal()] != null) {
      return writers[minorType.ordinal()];
    }
    switch (minorType) {
    case STRUCT:
      return getStructWriter();
    case LIST:
      return getListWriter();
    case TINYINT:
      return getTinyIntWriter();
    case UINT1:
      return getUInt1Writer();
    case UINT2:
      return getUInt2Writer();
    case SMALLINT:
      return getSmallIntWriter();
...
}

If a writer is not added to the writers array, the if statement evaluates to false, and the switch statement is run. During the process, the new writer is created and added to the array. So next time when the method is called, the if statements evaluates to true, and returns the corresponding writer immediately.

**Comment by emkornfield** [comment]
For setPosition, if I am reading correctly. Suppose there is only 2 writers.  Before the code looked like this when the loop is unrolled:

```
writers.get(0).setPosition(0);
writers.get(1).setPosition(0);
```

With the new code I believe it looks like this:
```
if (writers[0] != null) {
  writers[0].setPosition(0);
} 
if (writers[1] != null) {
  writes[1].setPosition(0);
} 
if (writers[2] != null) {
   writes[2].setPosition(0);
}
if (writers[3] != null) {
  writers[3].setPosition(0);
}
if (writers[4] != null) {
  writers[4].setPosition(0);
}   
...
if (writers[49] != null) {
  writers[4].setPosition(0);
}  
```

**Comment by liyafan82** [comment]
> For setPosition, if I am reading correctly. Suppose there is only 2 writers. Before the code looked like this when the loop is unrolled:
> 
> ```
> writers.get(0).setPosition(0);
> writers.get(1).setPosition(0);
> ```
> 
> With the new code I believe it looks like this:
> 
> ```
> if (writers[0] != null) {
>   writers[0].setPosition(0);
> } 
> if (writers[1] != null) {
>   writes[1].setPosition(0);
> } 
> if (writers[2] != null) {
>    writes[2].setPosition(0);
> }
> if (writers[3] != null) {
>   writers[3].setPosition(0);
> }
> if (writers[4] != null) {
>   writers[4].setPosition(0);
> }   
> ...
> if (writers[49] != null) {
>   writers[4].setPosition(0);
> }  
> ```

@emkornfield 
I see your good point. Thanks.
This is definitely going to have performance overhead.
I have revised the code to preserve the writer list. Please take a look.

**Comment by emkornfield** [comment]
@jacques-n any thoughts here?

**Comment by liyafan82** [comment]
@emkornfield @jacques-n dear sirs, would you please give another look at this?

Since it has been a long time since this PR was opened, please allow me to re-summarize it here. 

For UnionVector, a common operation is getting the underlying vector for each slot. All get/set methods for all data types involve this operation. 

We improve the performance of getting the underlying vector:

* The previous implementation: compare the type code against each minor type, and select the correct underlying vector.
* We save all underlying vectors into an array, which is indexed by the ordinal of the minor type, so we can find a vector directly given a type code.

IMO, this improvement is important, because it affects the performance of all key operations of UnionVector (all getXXX, setXXX methods), which will make code that uses UnionVector more efficient.

**Comment by jacques-n** [comment]
I think this change extends/builds on top of problems within the sparse union implementation. We should address that before adding more structures built on this. For now, I'm rejecting this change.

**Pull request closed without merging**